### ISSUE TITLE
Wrapping XML comments doesn't work in C++ code.

### ISSUE THREAD

**Issue by wansor** [stkb/Rewrap#199]
I'm using XML comments in C++. Alas, it doesn't always work. It works fine for a summary section and sometimes even for other sections. But too often it doesn't work as expected. It looks like wrapping stops when a param section has been found, like  

    <summary>wraps</summary>
    <param name="foo">doesn't wrap</param>
    <remarks>doesn't wrap</remarks>

Without the param section, the remark section will also wrap.

Expected behavior is wrapping all comment lines regardless of the current or previous sections. This might conflict with issue 94, but I installed Rewrap to get rid of comments running off the window, not only some of them.

**Comment by stkb** [comment]
This isn't related to #94 (which by the way, if you have an opinion on that issue, I'd appreciate it if you weigh-in on the proposal there or just +1, so I know if it's the right thing to do).

XML comments aren't currently supported in C++; it assumes you're writing Doxygen instead. I can change it so that `///` comments are assumed to be XML doc-comments (like for C#), and hope that that isn't a problem for anyone else (`//!` can also be used for Doxygen comments so hopefully it won't be).

**Comment by wansor** [comment]
Actually, I use XML comments to be processed by Doxygen. I prefer them over pure Doxygen comments, because Visual Studio (Intellisense) understands them and shows summaries or parameter descriptions while typing. Without this benefit, I would also prefer pure Doxygen comments without all that XML overhead.

I experimented a bit more and found out that XML attributes are causing my problems. Within a comment block, Rewrap works (more or less) as expected until it encounters an XML element with attribute(s). Then it stops working for the remainder of that comment block.

The expected behavior would be to continue working as if there had been no attributes. I do not expect (and wouldn't want) Rewrap to work differently depending on the name of an XML element, neither in C++ nor in C# code (that much against #94).

IMO, Rewrap should only distinguish between top level elements (like summary, param, etc.) and nested elements (like em, i, etc.), but not by their name but only their nesting level.

- For top level elements, their content should be wrapped.
- Nested elements should be treated as plain text.
- There should always be a line break before the opening tag of a top level element.
- There should always be a line break after the closing tag of a top level element.

Also, it would be nice to have t´some way to specify a hard line break that Rewrap wouldn't touch. Always ending a line after <br/>, </p> or </div> might be a solution but that contradicts my wish that nested elements should be treated as plain text. Keeping a line break if it is preceded by (at least) two spaces (like some markdown dialects do) might be an alterative.

-

**Comment by stkb** [comment]
Have now released v1.11.1 which contains a change. Before, all comments in C/C++ files were treated as markdown, with some extra rules to break before @tags, to support javadoc-style doxygen. This meant that xml tags were treated under the markdown rules for html tags, which wouldn't have been what you wanted.

I've now made a change so that `///` comments expect xml tags, so they now work the same as for .Net languages. This should work better for you.

The current rules for (all) xml tags are: it preserves newlines before & after (both start & end) tags.

So if you write it like this:

``` csharp
/// <summary>
/// Summary of what the function does
/// </summary>
```

It wraps like this:
``` csharp
/// <summary>
/// Summary of what
/// the function does
/// </summary>
```

But if you write it like this:

``` csharp
/// <summary>Summary of what the function does</summary>
```

It wraps like this:
``` csharp
/// <summary>Summary of what
/// the function does</summary>
```

---

I agree that inline elements should be treated differently (effectively "ignored") (see also #174) and I will work on that for the next version.

Also yes I think we can add the "two-spaces" (or single `\`) rule from markdown to preserve a line-break.

**Comment by wansor** [comment]
It works. Thanks a lot.

**Comment by stkb** [comment]
Have added a couple of changes to v1.12.0 (now released). Xml now supports the two-spaces to preserve a line-break, just like markdown.

It now also preserves the line breaks before/after certain "block" elements. I did however go with the tag name approach, with a list of known "block" tags, and all  others are treated as inline.

The nesting level-based approach is an interesting one though and I may change to it later, and it requires using a parser that's aware of where it is in the stack, but that's needed to fix #211 anyway.

However you do still need to look at tag names in some cases. For example the `<list>` element has child `<item>` elements, which also have block child elements. And the contents of `<code>` elements mustn't be wrapped. (Though maybe the comments in those code sections [#215])

I'll close this now.
